sparse_strips/vello_macros/src/lib.rs

```diff
@@ -0,0 +1,225 @@
+    transparent: bool,
+    /// Whether the test should not be run on the CPU (`vello_cpu`).
+    skip_cpu: bool,
+    /// Whether the test should not be run on the GPU (`vello_hybrid`).
+    skip_gpu: bool,
+    /// Whether no reference image should actually be created (for tests that only check
+    /// for panics, but are not interested in the actual output).
+    no_ref: bool,
+}
+
+impl Default for Arguments {
+    fn default() -> Self {
+        Self {
+            width: 100,
+            height: 100,
+            threshold: 0,
+            transparent: false,
+            skip_cpu: false,
+            skip_gpu: false,
+            no_ref: false,
+        }
+    }
+}
+
+/// Create a new vello snapshot test.
+#[proc_macro_attribute]
+pub fn v_test(attr: TokenStream, item: TokenStream) -> TokenStream {
+    let attrs = parse_macro_input!(attr as AttributeInput);
+
+    let input_fn = parse_macro_input!(item as ItemFn);
+
+    let input_fn_name = input_fn.sig.ident.clone();
+    let u8_fn_name = Ident::new(&format!("{}_cpu_u8", input_fn_name), input_fn_name.span());
+    let hybrid_fn_name = Ident::new(&format!("{}_gpu", input_fn_name), input_fn_name.span());
+
+    let input_fn_name_str = input_fn_name.to_string();
+    let u8_fn_name_str = u8_fn_name.to_string();
+    let hybrid_fn_name_str = hybrid_fn_name.to_string();
+
+    let Arguments {
+        width,
+        height,
+        threshold,
+        transparent,
+        skip_cpu,
+        mut skip_gpu,
+        no_ref,
+    } = parse_args(&attrs);
+
+    // These tests currently don't work with `vello_hybrid`.
+    skip_gpu |= {
+        input_fn_name_str.contains("clip")
+            || input_fn_name_str.contains("compose")
+            || input_fn_name_str.contains("gradient")
+            || input_fn_name_str.contains("image")
+            || input_fn_name_str.contains("layer")
+            || input_fn_name_str.contains("mask")
+            || input_fn_name_str.contains("mix")
+            || input_fn_name_str.contains("opacity")
+    };
+
+    let empty_snippet = quote! {};
+    let ignore_snippet = quote! {#[ignore]};
+
+    let ignore_hybrid = if skip_gpu {
+        ignore_snippet.clone()
+    } else {
+        empty_snippet.clone()
+    };
+    let ignore_cpu = if skip_cpu {
+        ignore_snippet.clone()
+    } else {
+        empty_snippet.clone()
+    };
+    let cpu_threshold = threshold;
+    let gpu_threshold = threshold + 1;
+
+    let expanded = quote! {
+        #input_fn
+
+        #ignore_cpu
+        #[test]
+        fn #u8_fn_name() {
+            use crate::util::{
+                check_ref, get_ctx
+            };
+            use vello_cpu::RenderContext;
+
+            let mut ctx = get_ctx::<RenderContext>(#width, #height, #transparent);
+            #input_fn_name(&mut ctx);
+            if !#no_ref {
+                check_ref(&ctx, #input_fn_name_str, #u8_fn_name_str, #cpu_threshold, true);
+            }
```

Comment:
Snapshot names can currently clash between test modules. I.e., in the below scenario, we would generate snapshots called `with_foo.png` for both modules. 

There is also inconsistent use of module name prefixing across tests (e.g. clip tests are prefixed with `clip_` but issues and basic are not). I think we can either choose a consistent naming pattern or prefix snapshots as below. I don't mind which!

```
mod clip {
   // ...
   fn with_foo() { ... }
}
mod gradient {
   // ...
   fn with_foo() { ... }
}
```




```suggestion
            if !#no_ref {
                let module_path = module_path!().strip_prefix("tests::").unwrap();
                let ref_name = format!("{}::{}", module_path, #input_fn_name_str);
                check_ref(&ctx, &ref_name, #u8_fn_name_str, #cpu_threshold, true);
            }
```

Reply:
You are right about that, I didn't know that you can access the module path like that! I think I will add this as a TODO for now because I don't want to rename all files (in this already very big PR) again, but I agree that we should change that.